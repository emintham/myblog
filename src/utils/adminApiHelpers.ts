// src/utils/adminApiHelpers.ts
import { dump } from 'js-yaml';
import type { PostApiPayload, FrontmatterObject } from '../types/admin';
import { AUTHOR_NAME } from '../siteConfig'; // Assuming AUTHOR_NAME is exported from siteConfig
import sharp from 'sharp';
import path from 'node:path';
import fs from 'node:fs/promises';

const projectRoot = process.cwd(); // Define projectRoot once
const originalImagesDir = path.join(projectRoot, 'images', 'originals'); // Define path to original images

/**
 * Transforms the raw API payload (similar to PostFormData) into a structured
 * FrontmatterObject suitable for writing to a file.
 * @param payload The raw data received by the API endpoint.
 * @returns A structured object for frontmatter.
 */
export async function transformApiPayloadToFrontmatter(payload: PostApiPayload): Promise<FrontmatterObject> {
  const frontmatter: FrontmatterObject = {
    title: payload.title,
    pubDate: new Date(payload.pubDate), // Convert string to Date object
    author: AUTHOR_NAME, // Set default author
    postType: payload.postType,
    draft: payload.draft === true || payload.draft === 'on' || (typeof payload.draft === 'string' && payload.draft.toLowerCase() === 'true'),
  };

  if (payload.description) {
    frontmatter.description = payload.description;
  }

  if (payload.tags) {
    const tagsArray = (typeof payload.tags === 'string' ? payload.tags.split(',') : Array.isArray(payload.tags) ? payload.tags : [])
      .map((tag: string) => tag.trim()).filter((tag: string) => tag);
    if (tagsArray.length > 0) {
      frontmatter.tags = tagsArray;
    }
  }

  if (payload.series) {
    frontmatter.series = payload.series;
  }

  // Book Note specific fields
  if (payload.postType === 'bookNote') {
    if (payload.bookTitle) frontmatter.bookTitle = payload.bookTitle;
    if (payload.bookAuthor) frontmatter.bookAuthor = payload.bookAuthor;

    let finalBookCover: { imageName?: string; alt?: string; originalWidth?: number } = {};
    // Check if bookCover is already an object (e.g., from update handler if it was already structured)
    if (payload.bookCover && typeof payload.bookCover === 'object' && (payload.bookCover.imageName || payload.bookCover.alt)) {
        finalBookCover = {
            imageName: payload.bookCover.imageName || '',
            alt: payload.bookCover.alt || '',
            originalWidth: payload.bookCover.originalWidth // Preserve if already there
        };
    } else if (payload.bookCoverImageName || payload.bookCoverAlt) { // From flat form fields
        finalBookCover = {
            imageName: payload.bookCoverImageName || '',
            alt: payload.bookCoverAlt || ''
        };
    }

    if (finalBookCover.imageName && !finalBookCover.originalWidth) { // Only try to get width if imageName exists and width isn't already set
      const imageBaseName = finalBookCover.imageName;
      const extensionsToTry = ['.jpg', '.jpeg', '.png', '.webp'];
      let foundImagePath: string | null = null;

      for (const ext of extensionsToTry) {
        const testPath = path.join(originalImagesDir, `${imageBaseName}${ext}`);
        try {
          await fs.access(testPath);
          foundImagePath = testPath;
          break;
        } catch {
          // File not found with this extension, try next
        }
      }

      if (foundImagePath) {
        try {
          const metadata = await sharp(foundImagePath).metadata();
          if (metadata.width) {
            finalBookCover.originalWidth = metadata.width;
          }
        } catch (sharpError) {
          console.warn(`[AdminAPIHelpers] Error getting metadata for ${foundImagePath} with sharp:`, sharpError);
        }
      } else {
        console.warn(`[AdminAPIHelpers] Original image not found for ${imageBaseName} in ${originalImagesDir}`);
      }
    }

    if (finalBookCover.imageName || finalBookCover.alt || finalBookCover.originalWidth) { // Only add if there's something to add
        frontmatter.bookCover = finalBookCover;
    }


    if (payload.quotesRef) frontmatter.quotesRef = payload.quotesRef;

    if (payload.bookTags) {
      const bookTagsArray = (typeof payload.bookTags === 'string' ? payload.bookTags.split(',') : Array.isArray(payload.bookTags) ? payload.bookTags : [])
        .map((tag: string) => tag.trim()).filter((tag: string) => tag);
      if (bookTagsArray.length > 0) {
        frontmatter.bookTags = bookTagsArray;
      }
    }
  }

  // Remove any top-level undefined properties that might have been set conditionally
  // Note: nested objects like bookCover are handled by their specific logic
  for (const key in frontmatter) {
    if (frontmatter[key as keyof FrontmatterObject] === undefined) {
      delete frontmatter[key as keyof FrontmatterObject];
    }
  }

  return frontmatter;
}

/**
 * Generates the full string content for a post file, including YAML frontmatter
 * and the body content.
 * @param frontmatter The structured frontmatter object.
 * @param bodyContent The main markdown body of the post.
 * @param postType The type of the post (e.g., "fleeting").
 * @param isNewPost Boolean flag, true if creating a new post, false if updating.
 * @returns A string representing the complete file content.
 */
export function generatePostFileContent(
  frontmatter: FrontmatterObject,
  bodyContent: string,
  postType: string,
  isNewPost: boolean
): string {
  const frontmatterString = dump(frontmatter, { skipInvalid: true, sortKeys: false }); // skipInvalid for safety, sortKeys: false to maintain order

  let fullContent = `---\n${frontmatterString}---\n\n`;

  const trimmedBody = (bodyContent || '').trim();

  if (trimmedBody) {
    fullContent += trimmedBody;
  } else if (isNewPost && postType !== 'fleeting') {
    // Add a placeholder for new, non-fleeting posts if body is empty
    fullContent += '## Start Writing Here\n\nReplace this with your content.';
  } else {
    // For fleeting posts or updates with an intentionally empty body, add nothing or just a newline
    fullContent += ''; // Or '\n' if a trailing newline is desired for empty bodies
  }
  // Ensure a final newline character for POSIX compatibility if not already ending with one
  if (!fullContent.endsWith('\n')) {
    fullContent += '\n';
  }

  return fullContent;
}