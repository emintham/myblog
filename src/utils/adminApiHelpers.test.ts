// src/utils/adminApiHelpers.test.ts
import { describe, it, expect } from 'vitest';
import { generatePostFileContent } from './adminApiHelpers';
import type { FrontmatterObject } from '../types/admin'; // Assuming this type is correctly exported

describe('generatePostFileContent', () => {
  const baseFrontmatter: FrontmatterObject = {
    title: 'Test Post',
    pubDate: new Date('2023-01-01T12:00:00Z'),
    author: 'Test Author',
    postType: 'blog',
    draft: false,
  };

  it('should generate content with frontmatter and body', () => {
    const bodyContent = 'This is the main content.';
    const result = generatePostFileContent(baseFrontmatter, bodyContent, 'blog', true);
    expect(result).toContain('title: Test Post');
    expect(result).toContain('pubDate: 2023-01-01T12:00:00.000Z'); // Default Date.toISOString() format
    expect(result).toContain('author: Test Author');
    expect(result).toContain('postType: blog');
    expect(result).toContain('draft: false');
    expect(result).toContain('\n---\n\nThis is the main content.\n');
  });

  it('should add placeholder for new, non-fleeting posts with empty body', () => {
    const result = generatePostFileContent(baseFrontmatter, '', 'blog', true);
    expect(result).toContain('\n---\n\n## Start Writing Here\n\nReplace this with your content.\n');
  });

  it('should not add placeholder for fleeting posts with empty body', () => {
    const fleetingFrontmatter = { ...baseFrontmatter, postType: 'fleeting' };
    const result = generatePostFileContent(fleetingFrontmatter, '', 'fleeting', true);
    // Expecting YAML frontmatter ending with \n--- and then two newlines for an empty body
    expect(result.endsWith('\n---\n\n')).toBe(true);
    expect(result).not.toContain('## Start Writing Here');
  });

  it('should not add placeholder when updating a post with empty body', () => {
    const result = generatePostFileContent(baseFrontmatter, '', 'blog', false); // isNewPost = false
    // Expecting YAML frontmatter ending with \n--- and then two newlines for an empty body
    expect(result.endsWith('\n---\n\n')).toBe(true);
    expect(result).not.toContain('## Start Writing Here');
  });

  it('should correctly format YAML frontmatter', () => {
    const result = generatePostFileContent(baseFrontmatter, 'Body', 'blog', true);
    const frontmatterPart = result.split('---')[1]; // Get content between ---
    expect(frontmatterPart).toMatchSnapshot(); // Using snapshot for complex structure
  });

  it('should ensure a final newline character', () => {
    const result = generatePostFileContent(baseFrontmatter, 'Some content', 'blog', true);
    expect(result.endsWith('\n')).toBe(true);
  });

  it('should ensure a final newline even if body is empty and no placeholder is added', () => {
    const result = generatePostFileContent(baseFrontmatter, '', 'blog', false); // isNewPost = false
    expect(result.endsWith('\n')).toBe(true);
  });

  it('should handle frontmatter with optional fields', () => {
    const fullFrontmatter: FrontmatterObject = {
      ...baseFrontmatter,
      description: 'A test description.',
      tags: ['test', 'sample'],
      series: 'Test Series',
    };
    const result = generatePostFileContent(fullFrontmatter, 'Body', 'blog', true);
    expect(result).toContain('description: A test description.');
    expect(result).toContain('tags:');
    expect(result).toContain('- test');
    expect(result).toContain('- sample');
    expect(result).toContain('series: Test Series');
    const frontmatterPart = result.split('---')[1];
    expect(frontmatterPart).toMatchSnapshot();
  });

  it('should handle empty body content correctly (trimmed)', () => {
    const bodyContent = '   '; // Only spaces
    const result = generatePostFileContent(baseFrontmatter, bodyContent, 'blog', false); // isNewPost = false (update)
    // Expecting that a body with only whitespace is treated as empty
    // So, YAML frontmatter ending with \n--- and then two newlines
    expect(result.endsWith('\n---\n\n')).toBe(true);
    expect(result.trim().endsWith('---')).toBe(true); // After trimming all newlines, should end with '---'
  });

  it('should correctly serialize date objects in frontmatter', () => {
    const date = new Date(2024, 0, 15, 10, 30, 0); // Jan 15, 2024, 10:30:00
    const fmWithDate: FrontmatterObject = { ...baseFrontmatter, pubDate: date };
    const result = generatePostFileContent(fmWithDate, 'Body', 'blog', true);
    // js-yaml typically serializes dates in ISO 8601 format, often including milliseconds
    expect(result).toContain(`pubDate: ${date.toISOString()}`);
  });
});
