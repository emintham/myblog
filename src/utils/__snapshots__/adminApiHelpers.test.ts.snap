// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generatePostFileContent > should correctly format YAML frontmatter 1`] = `
"
title: Test Post
pubDate: 2023-01-01T12:00:00.000Z
author: Test Author
postType: blog
draft: false
"
`;

exports[`generatePostFileContent > should handle frontmatter with optional fields 1`] = `
"
title: Test Post
pubDate: 2023-01-01T12:00:00.000Z
author: Test Author
postType: blog
draft: false
description: A test description.
tags:
  - test
  - sample
series: Test Series
"
`;
